sqlglot/expressions.py

```diff
@@ -6934,12 +6986,32 @@ def replace_children(expression: Expression, fun: t.Callable, *args, **kwargs) -
             if isinstance(cn, Expression):
                 for child_node in ensure_collection(fun(cn, *args, **kwargs)):
                     new_child_nodes.append(child_node)
-                    child_node.parent = expression
-                    child_node.arg_key = k
             else:
                 new_child_nodes.append(cn)
 
-        expression.args[k] = new_child_nodes if is_list_arg else seq_get(new_child_nodes, 0)
+        expression.set(k, new_child_nodes if is_list_arg else seq_get(new_child_nodes, 0))
+
+
+def replace_tree(expression: Expression, fun: t.Callable) -> Expression:
+    """
+    Replace an entire tree with the result of function calls on each node.
+
+    This will be traversed in reverse dfs, so leaves first.
+    If new nodes are created as a result of function calls, those will always be evaluated.
```

Comment:
Nice!

I caught myself thinking what "evaluated" means in this context, btw. May be worth rephrasing to clarify.